tools/toolchain/clang_lto_pgo.sh

```diff
@@ -0,0 +1,143 @@
+#!/bin/bash -uex
+
+if [ -z "${CLANG_ARCH}" ]; then
+    echo "CLANG_ARCH is not defined"
+elif [ "${CLANG_ARCH}" != "amd64" ]; then
+    LLVM_TARGET_ARCH=AArch64
+    LLVM_TARGET="-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=AArch64;WebAssembly"
+else
+    LLVM_TARGET_ARCH=X86
+    LLVM_TARGET="-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=WebAssembly"
+fi
+
+if [ -z "${CLANG_BUILD}" ]; then
+    echo "Skip building optimized clang"
+    exit 0
+elif [ "${CLANG_BUILD}" = "SKIP" ]; then
+    echo "Skip building optimized clang"
+    exit 0
+elif [ "${CLANG_BUILD}" = "BUILD" ]; then
+    echo "build optimized clang"
+elif [ "${CLANG_BUILD}" = "INSTALL" ]; then
+    echo "build and install optimized clang"
+elif [ "${CLANG_BUILD}" = "INSTALL_PREBUILT" ]; then
+    echo "install prebuilt optimized clang"
+else
+    echo "Not sure what to do with ${CLANG_BUILD}"
+    exit 1
+fi
+
+if [ ! -d "/optimized_clang" ]; then
+    echo "/optimized_clang not mounted"
+    exit 1
+fi
+DIR="/optimized_clang"
+
+STAGE0_BIN=$DIR/stage-0-${CLANG_ARCH}/build/bin
+STAGE1_BIN=$DIR/stage-1-${CLANG_ARCH}/build/bin
+
+# Which LLVM release to build in order to compile Scylla
+LLVM_SCYLLA_TAG=17.0.6
+
+# Which LLVM release to use to build clang (stage 0).
+LLVM_CLANG_TAG=17.0.6
+
+if [ "${CLANG_BUILD}" = "INSTALL" -o "${CLANG_BUILD}" = "BUILD" ]; then
+    cd "${DIR}"
+    # Clone, patch and bootstrap the newest Clang.
+    rm -rf stage-0-${CLANG_ARCH}
+    git clone https://github.com/llvm/llvm-project --branch llvmorg-${LLVM_CLANG_TAG} --depth=1 stage-0-${CLANG_ARCH}
+    cd stage-0-${CLANG_ARCH}
+    USE_CURRENT_COMPILER=(-DCMAKE_C_COMPILER="/usr/bin/clang" -DCMAKE_CXX_COMPILER="/usr/bin/clang++" -DLLVM_USE_LINKER="/usr/bin/ld")
+    COMMON_OPTS=(-DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD=all -DLLVM_TARGET_ARCH=${LLVM_TARGET_ARCH} -G Ninja -DLLVM_INCLUDE_BENCHMARKS=OFF -DLLVM_INCLUDE_EXAMPLES=OFF -DLLVM_INCLUDE_TESTS=OFF -DLLVM_ENABLE_BINDINGS=OFF)
+    SCYLLA_OPTS=(--date-stamp $(date "+%Y%m%d") --debuginfo 1 --tests-debuginfo 1 --c-compiler="$STAGE1_BIN/clang" --compiler="$STAGE1_BIN/clang++")
+
+    echo "stage-0: build the bootstrapped compiler"
+
+    cmake -B build -S llvm  "${USE_CURRENT_COMPILER[@]}" "${COMMON_OPTS[@]}" "${LLVM_TARGET}" -DLLVM_ENABLE_PROJECTS="clang;lld;bolt" -DLLVM_ENABLE_RUNTIMES="compiler-rt" -DCOMPILER_RT_BUILD_SANITIZERS=OFF -DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON
+    ninja -C build
+
+    USE_NEW_COMPILER=(-DCMAKE_C_COMPILER="$STAGE0_BIN/clang" -DCMAKE_CXX_COMPILER="$STAGE0_BIN/clang++" -DLLVM_USE_LINKER="$STAGE0_BIN/ld.lld")
+    NEW_COMMON_OPTS=(-DLLVM_ENABLE_PROJECTS="clang;lld" -DLLVM_ENABLE_RUNTIMES="compiler-rt" -DLLVM_ENABLE_LTO=Thin -DCLANG_DEFAULT_PIE_ON_LINUX=OFF -DLLVM_BUILD_TOOLS=OFF -DCMAKE_C_FLAGS="-Xclang -mllvm -Xclang -vp-counters-per-site=16" -DCMAKE_CXX_FLAGS="-Xclang -mllvm -Xclang -vp-counters-per-site=16")
+
```

Comment:
Note: I originally had to build the stage-0 LLVM because BOLT wasn't in Fedora packaging (and also because I needed some LLVM patches). But the situation has changed since then, so we don't really have to build the stage-0 anymore, we can just the system clang.

It doesn't make much of a difference, though.